IWDG timer

marnixcro opened the pull request:
**Summary:**
<!--
A summary, always referencing related issues:
Closes #0000, References #0000, etc.
-->

Add watchdog timer to freefall_lorawan. Closes #132.

<!--
Please motivate briefly why it is implemented this way, if that deviates
from the implementation proposal in the referenced issues.
-->

**Changes:**
<!-- What are the changes made in this pull request? -->

- BSP functions for IWDG
- Init and refresh functions in freefalllorawan (configurable in app_conf.h)

**Notes for Reviewers:**
<!--
How should your reviewers approach this pull request?
Any special requests or questions for specific reviewers?
-->

Set the watchdog timer on in app_conf.h before testing it.

I could also add IWDG to secure_element_lorawan and basic_lorawan.

marnixcro commented on Software/app/basic_lorawan/conf/app_conf.h:
```diff
@@ -51,6 +51,21 @@ typedef enum
+
+/* Watchdog counter trigger in seconds */
+/* This changes based on the config (prescalers for LSI and IWDG) */
+/* One tick is 1.024 seconds with default config: 1 / (LSI_VALUE / LSIPrescaler) = 1 / (250 / 256) */
```
Wrong calculations, will fix

elsalahy commented on Software/lib/GNSE_BSP/GNSE_bsp_clk_timer.c:
```diff
@@ -59,6 +60,59 @@ int32_t GNSE_BSP_RTC_Init(void)
+  */
+int32_t GNSE_BSP_IWDG_Refresh(void)
+{
+  if (HAL_IWDG_Refresh(&GNSE_BSP_iwdg) != HAL_OK)
```
Always returns ok, this is a simple register write?

elsalahy commented on Software/app/basic_lorawan/main.c:
```diff
@@ -47,6 +47,9 @@ int main(void)
   while (1)
   {
+#if defined(IWDG_TIMER_ON) && (IWDG_TIMER_ON == 1)
+    GNSE_BSP_IWDG_Refresh();
```
This will break functionality in case of sleep and any transmission duty cycle change

elsalahy commented on Software/lib/GNSE_BSP/GNSE_bsp_clk_timer.c:
```diff
@@ -59,6 +60,59 @@ int32_t GNSE_BSP_RTC_Init(void)
+  GNSE_BSP_iwdg.Instance = IWDG;
+  GNSE_BSP_iwdg.Init.Prescaler = IWDG_PRESCALER_256;
+  GNSE_BSP_iwdg.Init.Window = IWDG_WINR_WIN;
+  GNSE_BSP_iwdg.Init.Reload = GNSE_IWDG_TIMEOUT_S;
```
This can be based in the init, instead of assuming the app layer defines it

elsalahy commented on Software/target/stm32wlxx_hal_conf.h:
```diff
@@ -125,7 +125,7 @@
   */
 #if !defined (LSI_VALUE)
-#define LSI_VALUE                           32000UL     /*!< LSI Typical Value in Hz*/
+#define LSI_VALUE                           250UL       /*!< LSI Typical Value in Hz*/
```
As we discussed, it doesn't matter if you have max iwdg of 66 mins, it will need to be kicked in the end periodically by it's own function, or scheduled task, so the flexibility is good but when needed and now it's not needed

elsalahy commented on Software/lib/GNSE_BSP/GNSE_bsp_clk_timer.h:
```diff
@@ -26,9 +26,18 @@
+ * RCC_CSR_LSION takes some time to settle, causing an error in the clock config
+ * Only applicable when using a prescaler of 128 (check is only performed with this setting)
+ */
+#define IWDG_TIMER_DELAY 10U
```
why wait 10 specifically?

elsalahy commented on Software/lib/GNSE_BSP/GNSE_bsp_clk_timer.c:
```diff
@@ -59,6 +60,59 @@ int32_t GNSE_BSP_RTC_Init(void)
+  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_LSI;
+  RCC_OscInitStruct.LSIDiv = RCC_LSI_DIV128;
+  RCC_OscInitStruct.LSIState = RCC_LSI_ON;
+  HAL_Delay(IWDG_TIMER_DELAY); // Wait a bit for RCC_CSR_LSION to settle
```
The location of calling this function seems random, can you please  explain why it's called specifically here and not in another location like before tracer init?

elsalahy reviewed (changes requested):
The functionality seems to be ok 
1. Don’t go for max iwdg timeout, as it doesn’t matter, weather it’s small of big, at the end you will have to kick the dog (this is a term in Embedded :smile:) periodically

2. Schedule a task that resets the iwdg periodically
3. Adapt this to your freefall app first, and ignore basic_lorawan for now (I realized this after seeing your implementation that it will cause problems if we include it in any basic app)
4. IWDG timeout value is app specific, so it’s ok to pass it from the app layer to bsp via as an argument

marnixcro commented:
I've changed the watchdog timer to initiate a sequencer based on the reload value. I also moved it to freefall_lorawan. The IWDG is disabled by default.

elsalahy commented on Software/lib/GNSE_BSP/GNSE_bsp_clk_timer.c:
```diff
@@ -59,6 +70,55 @@ int32_t GNSE_BSP_RTC_Init(void)
+   * Change calculation when altering the window or prescaler settings
+   */
+  uint32_t IWDG_seq_timeout_ms = (GNSE_BSP_iwdg.Init.Reload / (LSI_VALUE / 256) * 1000);
+  UTIL_TIMER_Create(&IWDG_refresh_timer, 0xFFFFFFFFU, UTIL_TIMER_ONESHOT, GNSE_On_IWDG_Event, NULL);
+  UTIL_TIMER_SetPeriod(&IWDG_refresh_timer, IWDG_seq_timeout_ms);
+  UTIL_TIMER_Start(&IWDG_refresh_timer);
```
Please move scheduling to the application layer as the sequencer support is functional on the application layer (bare-metal)

Please adjust the code accordingly.

elsalahy reviewed (changes requested):
Looks good, scheduling is an application layer task and there is no BSP layer that supports scheduling.

elsalahy reviewed (approved):
Looks great, will be useful for other apps in the future
